teps/0139-trusted-artifacts.md

```diff
@@ -0,0 +1,477 @@
+[consumer : consume-content] /workspace/artifactStorage/afolder/c.txt
+```    
+</details>
+
+</br>
+
+
+The example pipeline shows a few things:
+- The solution of sharing content is generic, there's nothing in it specific to pipeline in question
+- Already in the pipeline the same code is used more than once. Replicating this solution in a pipeline with multiple producer or consumers would lead to a lot of duplication
+- The metadata (path, hash and type) required to trust an artifact is stored as a result in the status of the `TaskRun`. To complete the chain of trust we need to be able to trust the `status` of the `TaskRun`, feature that would be provided by the integration with Spire proposed in TEP-0089
+
+This suggests that we may be able to use a combination of API sugar-coating and controller injected steps to achieve the very same functionality while providing an API which is very similar to the one they are familiar with, but more powerful.
+
+### Goals
+
+- Contribute to the chain of trust by allowing consumer `Tasks` to trust artifacts on a `workspace` from producer `Tasks`, as long as results can be trusted. 
+
+### Non-Goals
+
+- This proposal is restricted to artifact in a `workspace`. There is no reason however that would prevent this mechanism to be extended to artifacts stored somewhere else. The same step injection mechanism could upload and download artifacts to and from other storage types (like OCI registries or object storage). This feature would be similar to what `PipelineResources` used to do, and it shall be designed in a separate TEP for the very reasons described in the [summary](#summary).
+- This proposal does not discuss how to expose artifacts outside of a pipeline, even though it sets foundations that could be used to achieve that
+- This proposal does not discuss how to inject artifact as inputs to a pipeline, even though it sets foundations that could be used to achieve that. For instance, one could use a workspace preprovisioned with artifacts and use artifact type params as inputs for a pipeline
+
+### Use Cases
+
+- Extend the chain of trust across Tasks for provenance produced by Tekton Chains based on the `TaskRuns` and `PipelineRuns` executed by Tekton Pipeline
+
+### Requirements
+
+Support the following combinations:
+
+- One producer, one consumer
+- One producer, N consumers, including with write-one, read-many storage class
+- Many producers, one consumer
+- Many producers, many consumers, including with write-one, read-many storage class
+- Fail validation if workspaces (static/runtime) are not fit for artifacts
+- Fail execution if checksum validation fail, surface error to TaskRun/PipelineRun failure reason
+
+## Proposal
+
+In short the proposal is to:
+
+- extend parameters and result types to a new type `artifact`, an object type with a fixed schema
+- extend the `Pod` logic in the controller to inject hashing and checking steps when required
```

Comment:
This is super cool! I think it aligns quite a bit with in-toto's artifact rules, and I wonder if it'd be possible to implement them here? Specifically, in-toto defines a MATCH rule that verifies artifact hashes match those recorded in the attestation for a prior step. Since this would be using data from Chains, I think it might make sense to unify the MATCH language.

https://github.com/in-toto/docs/blob/master/in-toto-spec.md#433-artifact-rules

Here's an example of using the rule with the provenance recorded for a prior step: https://github.com/adityasaky/in-toto-attestation-verifier/blob/main/layout-npm.yml#L34-L36

Reply:
Nice, thank you for sharing!

The API I was envisioning for Tekton was a list of artifacts, rather than match rules, since I need to be able to recalculate the digest/sha of specific artifacts to verify that they have not been compromised. 
We could let users specify MATCH rules on specific artifacts though, i.e. a certain folder artifact could be requested to match specific rules in the same format used by in-toto. Wdyt?

Reply:
Apologies as I'm unaware of the Tekton <-> Chains intricacies here. Would the artifact (+ hashes) list be distinct from Chains recording the same info for provenance? Will they be linked later, with this implementation providing the data to Chains?

> rather than match rules

Either way, the separation makes sense but I was thinking of it from the perspective of this snippet:

```
...
- name: aFileArtifact
  value: $(tasks.producer.results.aFileArtifact)
...
```

In my reading, artifacts are being listed, hashed, and matched against a prior record here, is that correct?

Reply:
Yes, the producer task also produces provenance data for its artifact, which is stored in the `TaskRun` status.
That provenance is made available by the Tekton controller to the consuming Task.